### PR TITLE
gh #44 Initial L3 C-Test Code generation.

### PR DESCRIPTION
**Pull request by srinivasgtl**
Issue #44 - Initial L3 low spec document for hdmi cec sink module.

### CODE REVIEW

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+### Pulse-Eight CEC Adaptor tool:
+The Pulse-Eight CEC Adapter will be used to frame and send commands to the DUT. This tool leverages libcec, which can be installed to facilitate CEC activities during testing. The cec-client tool will be employed extensively throughout the test. Additionally, RAFT can use this tool to automate the test cases.
+
+- Reference to the tool: https://www.pulse-eight.com/p/104/usb-hdmi-cec-adapter
+- libcec tools: https://www.pulse-eight.com/Download/Get/51  
+
+**Libraries and tools required for RAFT**
+- libcec should be installed on the PC where the Pulse-Eight CEC Adaptor is connected.
+- python-cec will be used to control the `CEC` adaptor.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:55]: python-cec should not be used to control the `CEC` adaptor.. Fixed yaml messages are required.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+```
+
+### Pulse-Eight CEC Adaptor tool:
+The Pulse-Eight CEC Adapter will be used to frame and send commands to the DUT. This tool leverages libcec, which can be installed to facilitate CEC activities during testing. The cec-client tool will be employed extensively throughout the test. Additionally, RAFT can use this tool to automate the test cases.
+
+- Reference to the tool: https://www.pulse-eight.com/p/104/usb-hdmi-cec-adapter
+- libcec tools: https://www.pulse-eight.com/Download/Get/51  
+
+**Libraries and tools required for RAFT**
+- libcec should be installed on the PC where the Pulse-Eight CEC Adaptor is connected.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:54]: libcec should not be required, because the messages are deterministic

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+- python-cec will be used to control the `CEC` adaptor.
+- https://pypi.org/project/cec/ will provide more information on how to use this library.
+  
+**cec-ctrl Commands used in Manual Mode**
+- libcec should be installed on the PC where the Pulse-Eight CEC Adaptor is connected.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:59]: libcec should not be installed.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+- To get the logical and physical address: `cec-client -s -d 4`
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:60]: you need to define a control message.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+- To Tx data: `cec-client -s -t 1 --data <command>` or `echo "tx <data>" | cec-client -s` Eg:`echo "tx 10:36" | cec-client -s`
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:61]: control library not valid for here.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+- Receive an HDMI OSD Command with a string of maximum length (14 bytes) from the CEC Adaptor and acknowledge it.
+- The CEC Adaptor sends a standby command to a device to all the logical addresses other than 0x00 and 0x0f.  `DUT` should not respond to any of these commands. 
+
+#### Message Broadcast and Receive broadcast 
+Note: All the below tests should be carried out on all the available HMDI ports.
+- Broadcast an HDMI CEC Command from the DUT and verify that this command has been received on the CEC Adaptor.
+- Receive a standby broadcasting command on the DUT sent by the CEC Adaptor and validate it.
+  
+#### Stress Test
+- Receiving an HDMI OSD Command repeatedly for 10 times with a different string of max length from CEC Adaptor.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:82]: that's a separate test and needs to be defined. with better reuqirements.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+- Transmit an HDMI OSD Command repeatedly for 10 times with a different string of max length to the CEC Adaptor.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:83]: needs more defination

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+#### Hardware Fault Test
+- Introducing a Fault on the HDMI line to test the Transmit functionality.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:86]: this doc needs to be min requirements for testing.

This hardware fault test desn't meet the requirements.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+# Test 1: Message Transmission and Reception Test - Unicast messages
+
+Functionality: 
+1. RAFT/user to send a simple "getCECVersion" command to the `DUT` before setting the logical address.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:91]: this is a control plane, message defined by a yaml file.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+2. `DUT` shall request a CEC Version from the CEC adaptor connected.  `DUT` should receive a valid CEC version.
+3. `DUT` shall receive an OSD Command with max buffer size from the CEC Adaptor and acknowledge to this command.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:93]: this is all very confused on what's expected here and doesn't define the process for how this will work.

Target :Trigger the menu test on L3, Wait for CEC Response.
Host : Send the yaml message to the system.
Target: Will output the message sent
Host: Will capture the message and compare the result against the expected result embedded in the yaml.

Repeat as required, define the list of yaml messages with their content, and expected result.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+| Title                         | Details                                          |
+|-------------------------------|--------------------------------------------------|
+| Function Name                 | `test_l3_hdmi_cec_sink_tx_rx_test`              |
+| Description                   | This test shall validate the Tx, and Rx CEC Commands between the `DUT`and CEC Adaptor connected on a network |
+| Test Group                    | 03                                               |
+| Test Case ID                  | 002                                              |
+| Priority                      | High                                             |
+
+**Pre-Conditions:**
+- The platforms are connected as shown in the picture where the CEC Adaptor is considered to be connected and detected correctly.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:105]: what configuration, ??

That needs to be defined by the yaml input files.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+- libcec is installed on the PC where the CEC Adaptor tool is connected. libcec shall respond to all the basic CEC Commands received from `DUT`
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:106]: - libcec is NOT expected to be installed.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+- RAFT shall provide a means to enter the LA for the test and any other data required during the test.
+- RAFT shall consider reading and give input to the test wherever required. 
+  
+**Sequence Diagram**
+```mermaid
+sequenceDiagram
+    participant DUT
+    participant CEC Adaptor
+    participant RAFT/user
+    RAFT/user->>CEC Adaptor: Read and validate LA and PA
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:129]: not valid.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+    RAFT/user->>DUT: RAFT/user to signal the test to move ahead
+    DUT->>DUT: Test fails if it has received any message.
+    RAFT/user->>RAFT/user: No CEC acknowledgment is received after CEC Frames are sent.   
+```
+
+#### Test Procedure 
+
+| Variation / Steps | Description | Test Data | Expected Result | Notes |
+|-------------------|-------------|-----------|------------------|-------|
+| 01 | Open HDMI CEC HAL using `HdmiCecOpen` API | `handle` = valid pointer | `HDMI_CEC_IO_SUCCESS` | Should be successful |
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:162]: Not valid.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+| 15 | Test to wait for the RAFT/user to signal about its test result for steps 13 and 14 | Press "Y"/"N" based on test result | N/A | RAFT/user to provide this test result |
+| 16 | Close HDMI CEC HAL using `HdmiCecClose` API | `handle` = valid handle | `HDMI_CEC_IO_SUCCESS` | Should be successful |
+
+# Test 2: Transmit and Receive CEC broadcast Commands 
+
+Functionality: 
+1. `DUT` shall broadcast a standby command to the connected devices and RAFT/user should validate this received command on the CEC Adaptor. 
+2. `DUT` shall receive a standby command as a broadcast command from the CEC Adaptor. RAFT/user to initiate and command the Test.
+
+| Title                         | Details                                          |
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:185]: needs re-doing.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+- RAFT shall initiate the commands to send the standby command.
+- RAFT shall consider reading and give input to the test wherever required. 
+    
+**Sequence Diagram**
+```mermaid
+sequenceDiagram
+    participant DUT
+    participant CEC Adaptor
+    participant RAFT/user
+    RAFT/user->>CEC Adaptor: Read and validate LA and PA (device ready)
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:216]: not valid.

File: profiles/sink/sink_hdmiCEC_test.yml [hunk 17]
```diff
@@ -0,0 +1,34 @@
+hdmicec:
+  config:
+    number_ports: 3
+    test:
+    tx:
+      #CEC GET Version command
+      - command:
```
Comment (Ulrond) [inline profiles/sink/sink_hdmiCEC_test.yml:7]: not correct as per the vDevice setup.

That's the reuiqrement.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+
+
+# Test 5: Hardware Fault Test
+
+Functionality: 
+1. `DUT` shall respond with the correct error code when a Fault (pulling the HDMI line to high) is introduced into the HDMI CEC line.
+
+| Title                         | Details                                          |
+|-------------------------------|--------------------------------------------------|
+| Function Name                 | `test_l3_hdmi_cec_sink_hw_fault_test`            |
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:323]: way down the line test and an edge case for much later in the project.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+| 05 | RAFT/user should receive this data and validate the command received. The test should wait for the RAFT to enter the test result based on the validation | N/A | N/A | RAFT to provide the result |
+| 06 | RAFT/user to command CEC adaptor to frame and broadcast the standby CEC command. Test data should be read from the test YAML file | `buffer` = test YAML file data | N/A | RAFT to set this data through CEC adaptor |
+| 07 | Wait for the RAFT to signal when the CEC broadcast is sent in step 6 | RAFT/user to provide `y` to move the test | N/A | RAFT to signal once the CEC broadcast is set |
+| 08 | Validate the data received from the CEC adaptor on the test RxCallback against the test YAML file | Test data received should be as mentioned in the test YAML file | Read and validate this data | Should be successful |
+| 09 | Close HDMI CEC HAL using `HdmiCecClose` API | `handle` = valid handle | `HDMI_CEC_IO_SUCCESS` | Should be successful |
+
+# Test 3: Stress Test
+
+Functionality: 
+1. `DUT` shall respond to a CEC OSD command received from the CEC Adaptor repeatedly for 10 times with different strings and validate.
```
Comment (Ulrond) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:249]: not enough information to figure out what should happen here.

Where's the yaml message that will work for both the vDevice & the platform code.

And the CEC adaptor is adnostic to the testing suite.

**Review by Ulrond — CHANGES_REQUESTED** [review]
this needs re-doing..

We need to have a meeting to discuss the high level requirement which this doesn't meet.


The tests are platform agnostic, and also will not know about the CEC Adaptor.

The test will run on both the vDevice and all platforms real or with CEC Adaptors, and must be platform adnostic and device Agnostic.

There will be no changes to the test to swap to vDevice, Platforms with CEC Adaptors or not.
The

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+- HDMI drivers should be up and running on the platform before running this test.
+- For now, it is suggested not to connect multiple CEC Adaptors onto a single PC as there is no proper support on libcec for multiple devices.
+
+```mermaid
+graph TB
+C[PC] <--> A
+A[Pulse-Eight CEC Adaptor ] <--> |HDMI| B[TV]  
+```
+
+### Pulse-Eight CEC Adaptor tool:
```
[the remark this answers is not in the document]
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:47]: We can remove the Pulse-Eight reference and mention as a CEC adaptor rather than saying "CEC device", else any person who is using this may raise a concern on the type of CEC device he needs to use.

When we say "any sort of CEC Device", what does it mean?  Do we have any scope for this?  Do we need to control any sort of Source devices, TV panels, AV Amplifiers, Repeaters, switches etc?

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md (hunk 2 above)
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:54]: libcec is the main software used to control the adaptor.  Else, we should specify what one needs to install here.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md (hunk 1 above)
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:55]: yaml can define only the CEC messages, but how will these messages sent from a device connected to the `DUT` or received from the `DUT`?  We already have a basic yaml file that defines the command.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md (hunk 4 above)
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:60]: there are few messages defined in the https://github.com/rdkcentral/rdk-halif-test-hdmi_cec/blob/feature/gh44_hdmicec_l3_sink_tests/profiles/sink/sink_hdmiCEC_test.yml.  But the CEC unicast commands should also use a logical address of the source device and we need to make sure that it uses the same address every time, else the test will fail.  Any idea on this?

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+  
+**Prerequisite Test to make sure all the CEC Adaptors are connected**
+1. Make sure to read the physical address of the CEC Adaptor allocated based on the port it has been connected to, using libCEC on PC.
+2. The physical and logical addresses allocated to the CEC adaptor shall be read and validated on RAFT.
+
+## Test Functionalities
+
+#### Message Transmission and Reception Test
+Note: All the below tests should be carried out on all the available HMDI ports.
+- Transmit an HDMI CEC basic command (GetCECVersion) from the DUT to receive a reply from the connected CEC Adaptor.
```
[the remark this answers is not in the document]
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:72]: GetCEC Version is a CEC Command that we need to frame and send to the connected device here and wait to get the reply from the connected device.  They are defined in the test yaml file.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+- Receive an HDMI OSD Command with a string of maximum length (14 bytes) from the CEC Adaptor and acknowledge it.
```
[the remark this answers is not in the document]
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:73]: RAFT Should ask to send this command from a source device that is connected to the `DUT`

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+- The CEC Adaptor sends a standby command to a device to all the logical addresses other than 0x00 and 0x0f.  `DUT` should not respond to any of these commands. 
+
+#### Message Broadcast and Receive broadcast 
+Note: All the below tests should be carried out on all the available HMDI ports.
+- Broadcast an HDMI CEC Command from the DUT and verify that this command has been received on the CEC Adaptor.
```
[the remark this answers is not in the document]
Comment (srinivasgtl) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:78]: Any API with a logical address with 0xF should be read as a broadcast command.  So when any device is connected to HDMI network receives these CEC frames should able to realize this address and consider the data.

**Review by srinivasgtl — COMMENTED** [review]
Let's discuss more on the design.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 24]
```diff
@@ -0,0 +1,346 @@
+# HDMI CEC L3 Test Document for Sink Devices.
+
+## Table of Contents
```
Comment (ishthiyaqahmed) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:3]: could you please list the overview as well in Table of Contents.

Comment (ishthiyaqahmed) [reply docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:3]: Do we have any references like high level test spec if so we need to include that as well.

File: docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md [hunk 25]
```diff
@@ -0,0 +1,346 @@
...
+
+The present document describes the test scope for the Sink Device activities only.
+
+### HDMI-CEC L3 Test Functionality
+The picture below depicts the HDMI CEC L3 Test Functionality Setup. This simple setup has an HDMI CEC Pulse-eight USB adaptor, which acts as a source device and can send the required commands (using libcec tool commands) and respond to the received CEC Commands.  
+
+Note: The below-shown prerequisites should be met before starting the test on the platforms. 
+- All the devices used in the test setup should support the `HDMI` `CEC` feature during the entire test duration.
+- HDMI drivers should be up and running on the platform before running this test.
+- For now, it is suggested not to connect multiple CEC Adaptors onto a single PC as there is no proper support on libcec for multiple devices.
```
Comment (ishthiyaqahmed) [inline docs/pages/hdmi-cec-sink-L3-Low-Level_TestSpec.md:39]: so currently we are testing on pc , is there any way to test on panels and set of boxes which supports CEC?

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        Initializes the test class with test-specific configurations.
+
+        Args:
+            log (logModule): Logging module instance to record test execution.
+        """
+        # Class variables
+        self.testName  = "test01_TransmitCECCommands"
+        self.qcID = '1'
+        self.tvLogicalAddress = '0'
```
Comment (srinivasgtl) [inline host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py:51]: Can we get this from config file as down the lane when we try to use the other logical address, the import from config file will hlep us

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+
+                # Transmit the CEC command
+                self.testhdmiCEC.cecTransmitCmd(destinationLogicalAddress, cec, payload)
+
+                time.sleep(2)
+
+                self.log.stepStart(f'HdmiCecTx Source: {deviceLogicalAddress} Destination: {destinationLogicalAddress} CEC OPCode: {cec} Payload: {payload}')
+
+                # Validate the transmission
+                result = self.hdmiCECController.checkTransmitStatus(deviceLogicalAddress, destinationLogicalAddress, cec, payload)
```
Comment (srinivasgtl) [inline host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py:117]: We need to validate two parameters here.  Staus and the result.  Also it differs between Tx and Broadcast.

**Review by srinivasgtl — COMMENTED** [review]
Please check for the return and status while tx.  return ack will differ between tx and broadcst.

File: src/test_l3_hdmi_cec_sink_driver.c [hunk 28]
```diff
@@ -0,0 +1,824 @@
...
+
+    // Building the CEC message
+    buf[0] = (uint8_t)((sourceLogicalAddress << 4) | (destinationLogicalAddress & 0x0F));
+    buf[1] = (uint8_t)cecCommand;
+    len = expectedDataLength + 2; // 1 byte for logical addresses, 1 byte for CEC Command
+
+    // Logging the transmission attempt
+    UT_LOG_INFO("Calling HdmiCecTx(IN:handle:[0x%0X], IN:buf:[%p], IN:len:[%d], OUT:result:[])", gHandle, buf, len);
+    int32_t status = HdmiCecTx(gHandle, buf, len, &result);
+    UT_LOG_INFO("Result HdmiCecTx(IN:handle:[0x%0X], IN:buf:[%p], IN:len:[%d], OUT:result:[%d]) HDMI_CEC_STATUS:[%s]", gHandle, buf, len, result, UT_Control_GetMapString(cecError_mapTable,status));
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply src/test_l3_hdmi_cec_sink_driver.c]: done, Added check for result

File: src/test_l3_hdmi_cec_sink_driver.c [hunk 29]
```diff
@@ -0,0 +1,824 @@
...
+void test_l3_hdmi_cec_sink_hal_RemoveLogicalAddress(void)
+{
+    gTestID = 7;
+    UT_LOG_INFO("In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
+
+    HDMI_CEC_STATUS status = HDMI_CEC_IO_SUCCESS;
+    int32_t logicalAddress = gLogicalAddress;
+    int32_t getLogicalAddress = 0;
+
+    assert(logicalAddress != -1);
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply src/test_l3_hdmi_cec_sink_driver.c:665]: this is in remove logical address. getlogical address after remove returns -1, but its not specified anywhere. So removed the check

File: src/test_l3_hdmi_cec_sink_driver.c [hunk 30]
```diff
@@ -0,0 +1,824 @@
...
+*
+* This test provides an interface to user/automation tool to transmit a CEC Command.
+* Necessary input should be provided to the test.
+*
+* **Test Group ID:** 02@n
+*
+* **Test Case ID:** 004@n
+*
+* **Pre-Conditions:** @n
+* HDMI-CEC Module should be intialized through Test 1 before calling this test.
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply src/test_l3_hdmi_cec_sink_driver.c:542]: Updated the comment

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py (hunk 26 above)
Comment (bhanucbp) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py]: Added a entry in yaml file and reading it now

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py (hunk 27 above)
Comment (bhanucbp) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py]: checkTransmitStatus reads the data bytes from the cec adaptor and validates, not from the device

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py [hunk 31]
```diff
@@ -0,0 +1,152 @@
...
+        5. Iterate through devices and send commands to appropriate logical addresses.
+        6. Validate transmission and record results.
+        7. Clean up by removing logical addresses and terminating the HDMI-CEC module.
+
+        Returns:
+            bool: Final result of the test execution.
+        """
+
+        # Initialize the hdmiCEC module
+        self.testhdmiCEC.initialise()
```
Comment (srinivasgtl) [inline host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py:87]: What if Init fails?  We are not validating this.  This is required to simulate a condition in source device where we try to connect more devices of same kind to check that init fails.

File: host/tests/classes/hdmiCEC.py [hunk 32]
```diff
@@ -0,0 +1,358 @@
...
+                   })
+
+        result = self.utMenu.select( self.testSuite, "Transmit CEC Command",promptWithAnswers)
+
+        pattern = r"Result HdmiCecTx\(.*OUT:result:\[(.*?)\]\).*HDMI_CEC_STATUS:\[(.*?)\]"
+
+        txStatus = self.searchPattern(result, pattern)
+
+        status = False
+        if txStatus == "HDMI_CEC_IO_SENT_AND_ACKD" or txStatus == "HDMI_CEC_IO_SENT_BUT_NOT_ACKD":
```
Comment (srinivasgtl) [inline host/tests/classes/hdmiCEC.py:237]: Should check for HDMI_CEC_IO_SENT_AND_ACKD when destination logical address is other than 0x0f (broadcast) for broadcast, it should be HDMI_CEC_IO_SENT_BUT_NOT_ACKD.  Should not check for either one of them.

File: host/tests/configs/example_rack_config.yml [hunk 33]
```diff
@@ -65,52 +56,30 @@ rackConfig:
                 # [ type: "ssh": port: 22 username: "test" password: "test" ]
                 # [ type: "telnet": port: 23 username: "test" password: "test" ]
                 - dut:
-                    ip: "127.0.0.1"  # IP Address of the ADA Hub
-                    description: "local PC"
-                    platform: "linux PC"
+                    ip: "127.0.0.1"
+                    description: "element"
```
Comment (srinivasgtl) [inline host/tests/configs/example_rack_config.yml:60]: while pushing this, make the platform platform as TV

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test02_ReceiveCECCommands.py [hunk 34]
```diff
@@ -0,0 +1,188 @@
...
+        3. Send commands to other devices and validate responses using callback data.
+        4. Validate response commands if specified.
+        5. Clean up by removing logical addresses and terminating the module.
+
+        Returns:
+            bool: Final result of the test execution.
+        """
+
+        # Initialize the hdmiCEC module
+        self.testhdmiCEC.initialise()
```
Comment (srinivasgtl) [inline host/tests/hdmiCEC_L3_Tests/hdmiCEC_test02_ReceiveCECCommands.py:101]: Check for the return status.

Comment (bhanucbp) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_test02_ReceiveCECCommands.py:101]: status check is already checked in c code

File: host/tests/configs/example_rack_config.yml (hunk 33 above)
Comment (bhanucbp) [reply host/tests/configs/example_rack_config.yml:60]: updated with tv

File: host/tests/classes/hdmiCEC.py (hunk 32 above)
Comment (bhanucbp) [reply host/tests/classes/hdmiCEC.py:237]: changed the condition check

**Review by Anbukannadhasan — APPROVED** [review]
Approved

File: src/test_l3_hdmi_cec_driver.c [hunk 35]
```diff
@@ -1,409 +1,809 @@
...
-* **Pre-Conditions:**  TODO: Add pre-conditions if any@n
-* **Dependencies:** TODO: Add dependencies if any@n
-*
-* Ref to API Definition specification documentation :[hdmi-cec_halSpec.md](../../docs/pages/hdmi-cec_halSpec.md)
-*/
+ * @defgroup HDMI_CEC_HALTESTS_L3 HDMI CEC HAL Tests L2 File
+ * @{
+ * @parblock
+ *
+ * ### L2 Test Cases for HDMI CEC HAL :
```
Comment (hari22yuva) [inline src/test_l3_hdmi_cec_driver.c:39]: remove L2 and L1 relegated the info in this file

File: docs/pages/hdmi-cec-L3_TestProcedure.md [hunk 36]
```diff
@@ -0,0 +1,366 @@
+# dsAudio HAL L3 Python Test Procedure
+
+## Table of Contents
```
Comment (hari22yuva) [inline docs/pages/hdmi-cec-L3_TestProcedure.md:3]: Update README.md

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py [hunk 37]
```diff
@@ -0,0 +1,152 @@
...
+
+        # Add the logical Address.
+        self.testhdmiCEC.addLogicalAddress(self.testLogicalAddress)
+
+        # Get the logical Address.
+        deviceLogicalAddress = self.testhdmiCEC.getLogicalAddress()
```
Comment (hari22yuva) [inline host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py:93]: can we check this function only for sink?

**Review by hari22yuva — COMMENTED** [review]
please check

File: host/tests/classes/hdmiCEC.py [hunk 38]
```diff
@@ -0,0 +1,362 @@
...
+            - If no match is found, None is returned.
+        """
+        match = re.search(pattern, haystack)
+        if match:
+            return match.group(1)
+        return None
+
+    def initialise(self):
+        """
+        Initializes the HDMI CEC module for sink.
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/classes/hdmiCEC.py:103]: removed

File: host/tests/classes/hdmiCEC.py [hunk 39]
```diff
@@ -0,0 +1,362 @@
...
+                "Initiator": initiator,
+                "Destination": destination,
+                "Data": parse_data_field(data)
+            })
+
+        return result
+
+    def getDeviceType(self):
+        """
+        Retrieves the type of the audio device.
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/classes/hdmiCEC.py:303]: removed

**Review by Ulrond — CHANGES_REQUESTED** [review]
- won't scale.
- unclear requirements for driver testing

File: src/test_l3_hdmi_cec_driver.c (hunk 35 above)
Comment (bhanucbp) [reply src/test_l3_hdmi_cec_driver.c:39]: done

File: docs/pages/hdmi-cec-L3_TestProcedure.md (hunk 36 above)
Comment (bhanucbp) [reply docs/pages/hdmi-cec-L3_TestProcedure.md:3]: updated

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py (hunk 37 above)
Comment (bhanucbp) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_test01_TransmitCECCommands.py:93]: check is already done in the c code

File: host/tests/configs/example_rack_config.yml [hunk 40]
```diff
@@ -65,52 +56,30 @@ rackConfig:
                 # [ type: "ssh": port: 22 username: "test" password: "test" ]
                 # [ type: "telnet": port: 23 username: "test" password: "test" ]
                 - dut:
-                    ip: "127.0.0.1"  # IP Address of the ADA Hub
-                    description: "local PC"
-                    platform: "linux PC"
+                    ip: "127.0.0.1"
+                    description: "tv"
+                    platform: "tv"
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/configs/example_rack_config.yml:61]: this is example configuration file, we update this file while testing

File: host/tests/configs/deviceConfig.yml [hunk 41]
```diff
@@ -24,5 +24,17 @@
 
 deviceConfig:
     cpe1:
-        platform:   "linux"
-        model:      "PC"
+        platform: "element"
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/configs/deviceConfig.yml]: this is example configuration file, we update this file while testing

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_L3_Runall.py [hunk 42]
```diff
@@ -0,0 +1,73 @@
...
+        try:
+            # Dynamically import the module
+            module = importlib.import_module(module_name)
+
+            # Dynamically access the test class from the module
+            # Assuming each test file has only one class named the same as the module
+            test_class = getattr(module, module_name)
+
+            # Instantiate and run the test
+            test_instance = test_class(hdmiCECSummerLog)
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_L3_Runall.py:66]: we will consider the messaging system in the next phase

File: profiles/sink/sink_hdmiCEC.yml [hunk 43]
```diff
@@ -4,3 +4,138 @@ hdmicec:
 
   features:
     extendedEnumsSupported: false
+
+  cec_responses:
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply profiles/sink/sink_hdmiCEC.yml:8]: moved responses to different yml file under assets folder

File: profiles/source/source_hdmiCEC.yml [hunk 44]
```diff
@@ -4,3 +4,138 @@ hdmicec:
 
   features:
     extendedEnumsSupported: false
+
+  cec_responses:
```
[the remark this answers is not in the document]
Comment (bhanucbp) [reply profiles/source/source_hdmiCEC.yml:8]: moved responses to different yml file under assets folder

File: host/tests/hdmiCEC_L3_Tests/hdmiCEC_L3_Runall.py (hunk 42 above)
[the remark this answers is not in the document]
Comment (Ulrond) [reply host/tests/hdmiCEC_L3_Tests/hdmiCEC_L3_Runall.py:66]: ignore me don't use the messaging system, we should just merging the changes for ConfigRead we'll do another thing for the messaging system.

**Review by Ulrond — APPROVED** [review]
- Later We should document the requirement on what the hdmi_cec_driver does
- We should also document each of the commands in more detail from the CEC, so we know what they are when we need to amend and maintain them

**Review by hari22yuva — APPROVED** [review]
approved as per requirement
